File: ios/eeuiGetui/eeuiGetuiAppModule.m
//
//  eeuiGetuiAppModule.m
//  Pods
//

#import "eeuiGetuiAppModule.h"
#import <WeexPluginLoader/WeexPluginLoader.h>
#import <GTSDK/GeTuiSdk.h>

@interface eeuiGetuiAppModule ()

@end

@implementation eeuiGetuiAppModule

@synthesize weexInstance;

WX_PlUGIN_EXPORT_MODULE(eeuiGetui, eeuiGetuiAppModule)
WX_EXPORT_METHOD(@selector(destroy))
WX_EXPORT_METHOD(@selector(stop))
WX_EXPORT_METHOD(@selector(resume))
WX_EXPORT_METHOD(@selector(turnOnPush))
WX_EXPORT_METHOD(@selector(turnOffPush))
WX_EXPORT_METHOD_SYNC(@selector(clientId))
WX_EXPORT_METHOD_SYNC(@selector(status))
WX_EXPORT_METHOD_SYNC(@selector(version))
WX_EXPORT_METHOD(@selector(runBackgroundEnable:))
WX_EXPORT_METHOD(@selector(lbsLocationEnable:andUserVerify:))
WX_EXPORT_METHOD(@selector(setChannelId:))
WX_EXPORT_METHOD(@selector(registerDeviceToken:))
WX_EXPORT_METHOD(@selector(bindAlias:andSequenceNum:))
WX_EXPORT_METHOD(@selector(unbindAlias:andSequenceNum:))
WX_EXPORT_METHOD(@selector(setTag:))
WX_EXPORT_METHOD(@selector(setPushModeForOff:))
WX_EXPORT_METHOD(@selector(setBadge:))
WX_EXPORT_METHOD(@selector(resetBadge))
WX_EXPORT_METHOD(@selector(sendMessage:error:))
WX_EXPORT_METHOD(@selector(sendFeedbackMessage:andTaskId:andMsgId:callback:))

/**
 *  销毁SDK，并且释放资源
 */

- (void)destroy {
    [GeTuiSdk destroy];
}

- (void)stop {
    [GeTuiSdk destroy];
}

/**
 *  恢复SDK运行,IOS7 以后支持Background Fetch方式，后台定期更新数据,该接口需要在Fetch起来后被调用，保证SDK 数据获取。
 */

- (void)resume {
    [GeTuiSdk resume];
}

/**
*   开启推送.
*/
- (void)turnOnPush {
    [GeTuiSdk setPushModeForOff:NO];
}

/**
*   关闭推送.
*/
- (void)turnOffPush {
    [GeTuiSdk setPushModeForOff:YES];
}

/**
 *  获取SDK的Cid
 *
 *  @return Cid值
 */
- (NSString *)clientId {
    return [GeTuiSdk clientId] ?: @"";
}

/**
 *  获取SDK运行状态
 *
 *  @return 运行状态
 */
- (int)status {
    return [GeTuiSdk status] == SdkStatusStarted ? 1 : 0;
}

/**
 *  获取SDK版本号
 *
 *  @return 版本值
 */
- (NSString *)version {
    return [GeTuiSdk version];
}

#pragma mark -

/**
 *  是否允许SDK 后台运行（默认值：NO）
 *  备注：可以未启动SDK就调用该方法
 *  警告：该功能会和音乐播放冲突，使用时请注意
 *
 *  @param isEnable 支持当APP进入后台后，个推是否运行,YES.允许
 */
- (void)runBackgroundEnable:(BOOL)isEnable {
    [GeTuiSdk runBackgroundEnable:isEnable];
}

/**
 *  地理围栏功能，设置地理围栏是否运行
 *  备注：SDK可以未启动就调用该方法
 *
 *  @param isEnable 设置地理围栏功能是否运行（默认值：NO）
 *  @param isVerify 设置是否SDK主动弹出用户定位请求（默认值：NO）
 */
- (void)lbsLocationEnable:(BOOL)isEnable andUserVerify:(BOOL)isVerify {
    [GeTuiSdk lbsLocationEnable:isEnable andUserVerify:isVerify];
}

#pragma mark -

/**
 *  设置渠道
 *  备注：SDK可以未启动就调用该方法
 *
 *  SDK-1.5.0+
 *
 *  @param aChannelId 渠道值，可以为空值
 */


- (void)setChannelId:(NSString *)aChannelId {
    [GeTuiSdk setChannelId:aChannelId];
}

/**
 *  向个推服务器注册DeviceToken
 *  备注：可以未启动SDK就调用该方法
 *
 *  @param deviceToken 推送时使用的deviceToken
 *
 */
- (void)registerDeviceToken:(NSString *)deviceToken {
    [GeTuiSdk registerDeviceToken:deviceToken];
}

/**
 *  绑定别名功能:后台可以根据别名进行推送
 *
 *  @param alias 别名字符串
 *  @param aSn   绑定序列码, 不为nil
 */
- (void)bindAlias:(NSString *)alias andSequenceNum:(NSString *)aSn {
    [GeTuiSdk bindAlias:alias andSequenceNum:aSn];
}

/**
 *  取消绑定别名功能
 *
 *  @param alias 别名字符串
 *  @param aSn   绑定序列码, 不为nil
 */
- (void)unbindAlias:(NSString *)alias andSequenceNum:(NSString *)aSn {
    [GeTuiSdk unbindAlias:alias andSequenceNum:aSn andIsSelf:YES];
}

/**
 *  给用户打标签 , 后台可以根据标签进行推送
 *
 *  @param tags 别名数组
 *
 *  @return 提交结果，YES表示尝试提交成功，NO表示尝试提交失败
 */
- (void)setTag:(NSArray *)tags {
    [GeTuiSdk setTags:tags];
}

/**
 *  设置关闭推送模式（默认值：NO）
 *
 *  @param isValue 消息推送开发，YES.关闭消息推送 NO.开启消息推送
 *
 *  SDK-1.2.1+
 *
 */
- (void)setPushModeForOff:(BOOL)isValue {
    [GeTuiSdk setPushModeForOff:isValue];
}


/**
 *  同步角标值到个推服务器
 *  该方法只是同步角标值到个推服务器，本地仍须调用setApplicationIconBadgeNumber函数
 *
 *  SDK-1.4.0+
 *
 *  @param value 角标数值
 */

- (void)setBadge:(NSUInteger)value {
    [GeTuiSdk setBadge:value];
}

- (void)resetBadge {
    [GeTuiSdk resetBadge];
}

#pragma mark -

/**
 *  SDK发送上行消息结果
 *
 *  @param body  需要发送的消息数据
 *  @param error 如果发送成功返回messageid，发送失败返回nil
 *
 *  @return 消息的msgId
 */
- (void)sendMessage:(NSData *)body error:(NSError **)error {
    [GeTuiSdk sendMessage:body error:error];
}

/**
 *  上行第三方自定义回执actionid
 *
 *  @param actionId 用户自定义的actionid，int类型，取值90001-90999。
 *  @param taskId   下发任务的任务ID
 *  @param msgId    下发任务的消息ID
 *
 *  @return BOOL，YES表示尝试提交成功，NO表示尝试提交失败。注：该结果不代表服务器收到该条数据
 *  该方法需要在回调方法“GeTuiSdkDidReceivePayload:andTaskId:andMessageId:andOffLine:fromApplication:”使用
 */
- (void)sendFeedbackMessage:(NSInteger)actionId andTaskId:(NSString *)taskId andMsgId:(NSString *)msgId callback:(WXModuleCallback)callback {
    BOOL isSuccess = [GeTuiSdk sendFeedbackMessage:actionId andTaskId:taskId andMsgId:msgId];
    callback(@[isSuccess ? @"true" : @"false"]);
}

@end

File: android/src/main/java/eeui/android/eeuiGetui/module/eeuiGetuiAppModule.java
package eeui.android.eeuiGetui.module;

import com.alibaba.fastjson.JSONArray;
import com.igexin.sdk.PushManager;
import com.igexin.sdk.Tag;
import com.taobao.weex.annotation.JSMethod;
import com.taobao.weex.bridge.JSCallback;

import app.eeui.framework.extend.base.WXModuleBase;
import app.eeui.framework.ui.eeui;

public class eeuiGetuiAppModule extends WXModuleBase {

    /**
     * Android 不存在 destroy方法，仅停止推送服务
     */
    @JSMethod
    public void destroy(){
        PushManager.getInstance().turnOffPush(eeui.getApplication());
    }

    /**
     * 停止SDK服务
     */
    @JSMethod
    public void stop(){
        PushManager.getInstance().turnOffPush(eeui.getApplication());
    }

    /**
     * 恢复SDK运行，重新接收推送
     */
    @JSMethod
    public void resume(){
        PushManager.getInstance().turnOnPush(eeui.getApplication());
    }

    /**
     * 打开SDK的推送.
     */
    @JSMethod
    public void turnOnPush(){
        PushManager.getInstance().turnOnPush(eeui.getApplication());
    }

    /**
     * 关闭SDK的推送.
     */
    @JSMethod
    public void turnOffPush(){
        PushManager.getInstance().turnOffPush(eeui.getApplication());
    }

    /**
     * 获取SDK的Cid
     *
     * @return Cid值
     */
    @JSMethod(uiThread = false)
    public String clientId(){
        return PushManager.getInstance().getClientid(eeui.getApplication());
    }

    /**
     * 获取SDK运行状态,
     *
     * @return 运行状态 1为已开启推送，0为已停止推送
     */
    @JSMethod(uiThread = false)
    public int status(){
        boolean isPushTurnOn = PushManager.getInstance().isPushTurnedOn(eeui.getApplication());
        return isPushTurnOn ? 1 : 0;
    }

    /**
     * 获取SDK版本号
     *
     * @return 版本号
     */
    @JSMethod(uiThread = false)
    public String version(){
        return PushManager.getInstance().getVersion(eeui.getApplication());
    }

    /**
     * 是否允许SDK 后台运行(默认为true)
     * 该方法在Android中无效，仅在iOS有效
     * @param isEnable
     */
    @JSMethod
    public void runBackgroundEnable(boolean isEnable){
        // Empty
    }

    /**
     * 地理围栏功能，设置地理围栏是否运行
     * 该方法在Android中无效，仅在iOS有效，在AndroidManifest.xml开启相应地权限
     * <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
     * <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
     * @param isEnable 设置地理围栏功能是否运行（默认值：NO）
     * @param isVerify 设置是否SDK主动弹出用户定位请求（默认值：NO）
     */
    @JSMethod
    public void lbsLocationEnable(boolean isEnable, boolean isVerify){
        // Empty
    }

    /**
     *  设置渠道
     *  该方法在Android中无效，仅在iOS有效Ø
     *  @param channelId 渠道值，可以为空值
     */
    @JSMethod
    public void setChannelId(String channelId){
        // Empty
    }

    /**
     * 向个推服务器注册DeviceToken
     * 该方法在Android中无效，仅在iOS有效
     * @param deviceToken
     */
    @JSMethod
    public void registerDeviceToken(String deviceToken){
        // Empty

    }

    /**
     * 绑定别名功能:后台可以根据别名进行推送
     *
     * @param alias 别名字符串
     * @param aSn   绑定序列码, Android中无效，仅在iOS有效
     */
    @JSMethod
    public void bindAlias(String alias, String aSn){
        PushManager.getInstance().bindAlias(eeui.getApplication(), alias);
    }

    /**
     *  取消绑定别名功能
     *
     *  @param alias 别名字符串
     *  @param aSn   绑定序列码, Android中无效，仅在iOS有效
     */
    @JSMethod
    public void unbindAlias(String alias, String aSn){
        PushManager.getInstance().unBindAlias(eeui.getApplication(), alias, false);
    }

    /**
     *  给用户打标签 , 后台可以根据标签进行推送
     *
     *  @param tags 别名数组
     */
    @JSMethod
    public void setTag(JSONArray tags){
        if (tags == null || tags.size() == 0){
            return;
        }

        Tag[] tagArray = new Tag[tags.size()];
        for (int i = 0; i < tags.size(); i++ ){
            Tag tag = new Tag();
            tag.setName(tags.getString(i));
            tagArray[i] = tag;
        }

        PushManager.getInstance().setTag(eeui.getApplication(), tagArray, "setTag");
    }

    /**
     * 设置关闭推送模式
     * Android中无效，仅在iOS有效
     *
     * @param isValue
     */
    @JSMethod
    public void setPushModeForOff(boolean isValue){
        // Empty
    }

    /**
     * 同步角标值到个推服务器
     * Android中无效，仅在iOS有效
     * @param value
     */
    @JSMethod
    public void setBadge(int value){
        // Empty
    }

    /**
     * 重置角标值到个推服务器
     * Android中无效，仅在iOS有效
     * @param badge
     */
    @JSMethod
    public void resetBadge(int badge){
        //Empty
    }


    /**
     *  SDK发送上行消息结果
     *  Android中无效，仅在iOS有效
     *  @param body  需要发送的消息数据
     *  @param error 如果发送成功返回messageid
     *
     */
    @JSMethod
    public void sendMessage(String body, String error){
        // Empty

    }

    /**
     *  上行第三方自定义回执actionid
     *  @param actionId 用户自定义的actionid，int类型，取值90001-90999。
     *  @param taskId   下发任务的任务ID
     *  @param msgId    下发任务的消息ID
     *
     *  @return 上行成功或失败，若上行失败，可能上行失败；taskid为空或者 messageid 为空 或者 actionid 不在取值范围以内
     */
    @JSMethod
    public void sendFeedbackMessage(int actionId, String taskId, String msgId, JSCallback callback){
        callback.invoke(PushManager.getInstance().sendFeedbackMessage(eeui.getApplication(), taskId, msgId, actionId));
    }

    /**
     * 设置静默时间
     *
     * @param beginHour 开始时间，设置范围在0-23小时之间，单位 h
     * @param duration 持续时间，设置范围在0-23小时之间。持续时间为0则不静默，单位 h
     */
    @JSMethod
    public void sendSilentTime(int beginHour, int duration, JSCallback callback){
        callback.invoke(PushManager.getInstance().setSilentTime(eeui.getApplication(), beginHour, duration));
    }

    /**
     * 设置Socket超时时间
     * @param times 超时时间
     */
    @JSMethod
    public void setSocketTimeout(int times){
        PushManager.getInstance().setSocketTimeout(eeui.getApplication(), times);
    }
}
